niceman/distributions/base.py

```diff
@@ -40,6 +40,108 @@ def TypedList(type_):
 #
 
 class SpecObject(object):
+
+    # Empty tuple would signal that all attribs will be used while establishing
+    # the "identity" of the specobject
+    _cmp_fields = tuple()
+    # Fields of the primary interest when showing diff
+    _diff_fields = tuple()
+
+    @property
+    def _cmp_id(self):
+        if not self._cmp_fields:
+            # Might need to be gone or some custom exception
+            raise RuntimeError(
+                "Cannot establish identity of %r since _cmp_fields "
+                "are not defined", self)
+        fields = self._cmp_fields \
+            if self._cmp_fields \
+            else self._attr_names
+        return tuple(getattr(self, a) for a in fields)
+
+    # TODO: make it "lazy" or may be there is already a helper in attrs?
+    @property
+    def _attr_names(self):
+        return tuple(f.name for f in self.__attrs_attrs__)
+
+    # TODO: code here is too generic and could be used for "is_identical_to"
+    #   kind of check, just logic could be adjusted
+    def is_satisfied_by(self, other):
+        """does the other spec object satisfies the requirements of the current one
+        """
+        # TODO(OPT): might explicitly set to no op function if lgr level is too high
+        log_cmp = lambda *args: lgr.log(5, *args)
+        log_cmp("Comparing (satisfy) self=%r to other=%r", self, other)
+        if not isinstance(other, self.__class__):
+            log_cmp("is not a subclass")
+            return False
+
+        # ??? probably not appropriate for 'satisfy' since e.g. pkg could be
+        #     missing arch spec and thus be satisfied
+        # if self._cmp_id != other._cmp_id:
+        #     log_cmp("%r is different from %r since cmd_id differ",
+        #             self._cmp_id, other._cmp_id)
+        #     return False
+
+        # TODO: any benefit from using ordered set may be to fail earlier etc?
+        # recently added OrderedSet to nibabel -- borrow from there
+        self_attrs = set(self._attr_names)
+        other_attrs = set(other._attr_names)
+        all_attrs = self_attrs.union(other_attrs)
+
+        # if `other` has some field set which is not set in `self` - that is Ok
+        # although (TODO) think about files!? (probably we would need custom
+        # satisfaction check for some fields like `files`)
```

Comment:
I'm not following the files bit.  If, say, a conda package was retraced with a particular file input, why should it matter?  Doesn't it just matter that the package is satisfied? I have version N of package X, then that file should be there.

Reply:
I agree.  I'm happy to wait for a real case to show itself before trying to figure this one out.